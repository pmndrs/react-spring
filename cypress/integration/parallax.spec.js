describe('Parallax - vertical', () => {
  const HEIGHT = Cypress.config('viewportHeight')

  beforeEach(() => {
    cy.visit('/vertical')
  })

  it('should translate layers as expected', () => {
    // intial snapshot
    cy.findByTestId('container').toMatchSnapshot()

    // initial layer positions
    cy.findByTestId('default-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(0px, ${HEIGHT * 2}px, 0px)`)
      )

    cy.findByTestId('opposite-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(${HEIGHT * 2}px, 0px, 0px)`)
      )

    cy.findByTestId('sticky-layer')
      .should('have.css', 'position', 'absolute')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(0px, ${HEIGHT}px, 0px)`)
      )

    // scroll to next page
    cy.findByTestId('container').scrollTo(0, HEIGHT)
    // wait for animation to finish
    cy.wait(2000)

    // halfway snapshot
    cy.findByTestId('container').toMatchSnapshot()

    // halfway layer positions
    cy.findByTestId('default-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(0px, ${HEIGHT}px, 0px)`)
      )

    cy.findByTestId('opposite-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(${HEIGHT}px, 0px, 0px)`)
      )

    cy.findByTestId('sticky-layer')
      .should('have.css', 'position', 'sticky')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(0px, 0px, 0px)`)
      )

    // scroll to last page
    cy.findByTestId('container').scrollTo('bottom')
    // wait for animation again
    cy.wait(2000)

    // final snapshot
    cy.findByTestId('container').toMatchSnapshot()

    // final layer positions
    cy.findAllByTestId(/-layer/).each(layer =>
      expect(layer[0].style.transform).to.equal('translate3d(0px, 0px, 0px)')
    )

    cy.findByTestId('sticky-layer').should('have.css', 'position', 'sticky')
  })

  it('should scroll to the correct page with scrollTo', () => {
    cy.findByRole('button').click()

    cy.findByTestId('container').invoke('scrollTop').should('equal', HEIGHT)
  })
})

describe('Parallax - horizontal', () => {
  const WIDTH = Cypress.config('viewportWidth')

  beforeEach(() => {
    cy.visit('/horizontal')
  })

  it('should translate layers as expected', () => {
    // intial snapshot
    cy.findByTestId('container').toMatchSnapshot()

    // initial layer positions
    cy.findByTestId('default-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(${WIDTH * 2}px, 0px, 0px)`)
      )

    cy.findByTestId('opposite-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(0px, ${WIDTH * 2}px, 0px)`)
      )

    cy.findByTestId('sticky-layer')
      .should('have.css', 'position', 'absolute')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(${WIDTH}px, 0px, 0px)`)
      )

    // scroll to next page
    cy.findByTestId('container').scrollTo(WIDTH, 0)
    // wait for animation to finish
    cy.wait(2000)

    // halfway snapshot
    cy.findByTestId('container').toMatchSnapshot()

    // halfway layer positions
    cy.findByTestId('default-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(${WIDTH}px, 0px, 0px)`)
      )

    cy.findByTestId('opposite-layer')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(0px, ${WIDTH}px, 0px)`)
      )

    cy.findByTestId('sticky-layer')
      .should('have.css', 'position', 'sticky')
      .then(layer => layer[0].style.transform)
      .then(transform =>
        expect(transform).to.equal(`translate3d(0px, 0px, 0px)`)
      )

    // a possible bug in cypress causes this to fail with `yarn cypress` (`cypress open`)
    // but will pass with `yarn test` (`cypress run`)
    cy.findByTestId('container').scrollTo('right')
    // wait for animation again
    cy.wait(2000)

    // final snapshot
    cy.findByTestId('container').toMatchSnapshot()

    // final layer positions
    cy.findAllByTestId(/-layer/).each(layer =>
      expect(layer[0].style.transform).to.equal('translate3d(0px, 0px, 0px)')
    )

    cy.findByTestId('sticky-layer').should('have.css', 'position', 'sticky')
  })

  it('should scroll to the correct page with scrollTo', () => {
    cy.findByRole('button').click()

    cy.findByTestId('container').invoke('scrollLeft').should('equal', WIDTH)
  })
})
