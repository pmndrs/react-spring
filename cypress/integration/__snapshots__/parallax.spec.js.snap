exports[`Parallax - vertical > should translate layers as expected #0`] = `
<div
  data-testid="container"
  class="container"
  style="position: absolute; width: 100%; height: 100%; overflow: hidden scroll; transform: translate3d(0px, 0px, 0px);"
></div>
`;

exports[`Parallax - vertical > should translate layers as expected #1`] = `
<div
  data-testid="container"
  class="container"
  style="position: absolute; width: 100%; height: 100%; overflow: hidden scroll; transform: translate3d(0px, 0px, 0px);"
>
  <div
    style="overflow: hidden; position: absolute; width: 100%; height: 1980px; transform: translate3d(0px, 0px, 0px);"
  >
    <div
      data-testid="opposite-layer"
      class="opposite"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 660px; transform: translate3d(660px, 0px, 0px);"
    ></div>
    <div
      data-testid="default-layer"
      class="default"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 660px; transform: translate3d(0px, 660px, 0px);"
    ></div>
    <div
      class="flex-center"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 660px; transform: translate3d(0px, 0px, 0px);"
    >
      <button>Scroll</button>
    </div>
  </div>
  <div
    data-testid="sticky-layer"
    class="sticky"
    style="position: sticky; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 660px; transform: translate3d(0px, 0px, 0px);"
  >
    <div class="sticky-content flex-center">Sticky</div>
  </div>
</div>
`;

exports[`Parallax - vertical > should translate layers as expected #2`] = `
<div
  data-testid="container"
  class="container"
  style="position: absolute; width: 100%; height: 100%; overflow: hidden scroll; transform: translate3d(0px, 0px, 0px);"
>
  <div
    style="overflow: hidden; position: absolute; width: 100%; height: 1980px; transform: translate3d(0px, 0px, 0px);"
  >
    <div
      data-testid="opposite-layer"
      class="opposite"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 660px; transform: translate3d(0px, 0px, 0px);"
    ></div>
    <div
      data-testid="default-layer"
      class="default"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 660px; transform: translate3d(0px, 0px, 0px);"
    ></div>
    <div
      class="flex-center"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 660px; transform: translate3d(0px, 0px, 0px);"
    >
      <button>Scroll</button>
    </div>
  </div>
  <div
    data-testid="sticky-layer"
    class="sticky"
    style="position: sticky; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 660px; transform: translate3d(0px, 0px, 0px);"
  >
    <div class="sticky-content flex-center">Sticky</div>
  </div>
</div>
`;

exports[`Parallax - horizontal > should translate layers as expected #0`] = `
<div
  data-testid="container"
  class="container"
  style="position: absolute; width: 100%; height: 100%; overflow: scroll hidden; transform: translate3d(0px, 0px, 0px);"
></div>
`;

exports[`Parallax - horizontal > should translate layers as expected #1`] = `
<div
  data-testid="container"
  class="container"
  style="position: absolute; width: 100%; height: 100%; overflow: scroll hidden; transform: translate3d(0px, 0px, 0px);"
>
  <div
    style="overflow: hidden; position: absolute; height: 100%; width: 3000px; transform: translate3d(0px, 0px, 0px);"
  >
    <div
      data-testid="opposite-layer"
      class="opposite"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 1000px; transform: translate3d(0px, 1000px, 0px);"
    ></div>
    <div
      data-testid="default-layer"
      class="default"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 1000px; transform: translate3d(1000px, 0px, 0px);"
    ></div>
    <div
      class="flex-center"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 1000px; transform: translate3d(0px, 0px, 0px);"
    >
      <button>Scroll</button>
    </div>
  </div>
  <div
    data-testid="sticky-layer"
    class="sticky"
    style="position: sticky; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 1000px; transform: translate3d(0px, 0px, 0px);"
  >
    <div class="sticky-content flex-center">Sticky</div>
  </div>
</div>
`;

exports[`Parallax - horizontal > should translate layers as expected #2`] = `
<div
  data-testid="container"
  class="container"
  style="position: absolute; width: 100%; height: 100%; overflow: scroll hidden; transform: translate3d(0px, 0px, 0px);"
>
  <div
    style="overflow: hidden; position: absolute; height: 100%; width: 3000px; transform: translate3d(0px, 0px, 0px);"
  >
    <div
      data-testid="opposite-layer"
      class="opposite"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; width: 100%; height: 1000px; transform: translate3d(0px, 0px, 0px);"
    ></div>
    <div
      data-testid="default-layer"
      class="default"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 1000px; transform: translate3d(0px, 0px, 0px);"
    ></div>
    <div
      class="flex-center"
      style="position: absolute; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 1000px; transform: translate3d(0px, 0px, 0px);"
    >
      <button>Scroll</button>
    </div>
  </div>
  <div
    data-testid="sticky-layer"
    class="sticky"
    style="position: sticky; inset: 0px; background-size: auto; background-repeat: no-repeat; will-change: transform; height: 100%; width: 1000px; transform: translate3d(0px, 0px, 0px);"
  >
    <div class="sticky-content flex-center">Sticky</div>
  </div>
</div>
`;
